### ISSUE TITLE
DRC clean designs drop contacts on GDS export

### ISSUE THREAD

**Issue by d-m-bailey** [RTimothyEdwards/magic#247]
On one of the [mpw-3 analog designs](https://foss-eda-tools.googlesource.com/third_party/shuttle/sky130/mpw-003/slot-016/), the tapeout data is [missing licon1](https://github.com/d-m-bailey/extra_be_checks/discussions/27#discussioncomment-5905914) that is visible in magic. 
<img width="841" alt="pcon" src="https://github.com/RTimothyEdwards/magic/assets/8349634/0b35f5ef-6a68-4034-a300-0978e20384af">
Note: cell has been rotated.
<img width="966" alt="no-licon1" src="https://github.com/RTimothyEdwards/magic/assets/8349634/c94fd3ba-0b19-4a8a-a8cd-51fd5952f30c">

The gds feedback shows 1816 errors
```
      8 feedback add "CIF error in cell Class_D_post, layer CONT: no room for contacts in area!" pale
   1800 feedback add "CIF error in cell Class_D_post, layer MCON: no room for contacts in area!" pale
      8 feedback add "CIF error in cell Class_D_post, layer NPC: no room for contacts in area!" pale
```
but there were no drc errors indicating that there would be a problem. Magic `8.3.400`

Maybe this should be a fatal error that produces no gds by default.

**Comment by RTimothyEdwards** [comment]
@d-m-bailey :  It's problematic to force that to be a fatal error because there are situations where it crops up on complicated polygon geometry---some areas are too small to fit contacts but it is not an issue for the overall region.  But the "+" shape above has caused the issue in a way that is not detected as a DRC error.  There is probably a way to detect that kind of issue;  it's a pathological case.  Maybe a count of vias generated in any single connected area, such that if the count is zero then a fatal error occurs.  Either than, or in the above case, the via generation routine needs to be able to deal with that pathological case.  At any rate, that case is one that I wasn't aware could happen.

**Comment by d-m-bailey** [comment]
@RTimothyEdwards Thanks for checking. 
So _most_ of the time, the errors are related to something like this
<img width="869" alt="MCON" src="https://github.com/RTimothyEdwards/magic/assets/8349634/7bb81f1a-dfec-4eef-a272-9c1ccc70a914">
where some vias are generated and some aren't. In this case too, the user might want to be know that only one via was created where they were expecting 3.

A quick and dirty solution might be to add a option `cif allowerrors [yes|no]` with a default of `no`.
At the end of `cif write <file>`, 
if errors are detected and `allowerrors` is `no`, 
- delete `<file>`
- display `<file> not created due to cif write errors. Use "cif allowerrors yes" to create <file> in spite of errors.`

if errors are detected and `allowerrors` is `yes`
- display `Warning: <file> created in spite of errors.`

**Comment by RTimothyEdwards** [comment]
@d-m-bailey :  Exactly.  The cut filling algorithm will start with a contact tile and then try to expand it in all directions to find the largest area available to place cuts in.  What makes the example that you posted up top so pathological is that the "+" shape can be expanded two ways, either up-down or side-to-side, so both ways would need to be checked to discover that a cut can only fit in one of those expansions.

**Comment by RTimothyEdwards** [comment]
Your example directly above is _not_ such a case, and is not a situation in which the user should expect that there will be three contacts---the area of the contact is only slightly bigger than a via.  The main issue here is that any contact area that is not flagging a DRC error should have at least one cut, and the pathological case is one instance (and possibly the only instance?) in which no DRC error occurs but no cut is placed.
